Add dict(*iterable) constructor

guilhermeleobas opened the pull request:
Fix issue #8328

esc commented on numba/typed/dictimpl.py:
```diff
@@ -1,9 +1,32 @@
+
+
+@lower_builtin(dict, types.IterableType)
+def list_constructor(context, builder, sig, args):
```
```suggestion
def dict_constructor(context, builder, sig, args):
```

esc commented on numba/typed/dictimpl.py:
```diff
@@ -1,9 +1,32 @@
+    dicttype = sig.return_type
+    kt, vt = dicttype.key_type, dicttype.value_type
+
+    def list_impl(iterable):
```
```suggestion
    def dict_impl(iterable):
```

esc commented on numba/typed/dictimpl.py:
```diff
@@ -1,9 +1,32 @@
+            res[k] = v
+        return res
+
+    return context.compile_internal(builder, list_impl, sig, args)
```
```suggestion
    return context.compile_internal(builder, dict_impl, sig, args)
```

guilhermeleobas replied:
Fixed!

esc reviewed (changes requested):
I gave this a first pass and found some potential copy and paste errors.

stuartarchibald commented on numba/core/typing/dictdecl.py:
```diff
@@ -17,13 +17,26 @@
 _message_dict_support = """
 Unsupported use of `dict()` with positional or keyword argument(s). \
-The only supported use is `dict()`.
+The only supported uses are `dict()` or `dict(*iterable)`.
```
```suggestion
The only supported uses are `dict()` or `dict(iterable)`.
```

stuartarchibald commented on docs/source/reference/pysupported.rst:
```diff
@@ -724,10 +724,10 @@ Typed Dict
+Numba supports the use of ``dict()``.  Such use is semantically equivalent to
+``{}`` and ``numba.typed.Dict()``. It will create an instance of
+``numba.typed.Dict`` where the key-value types will be later inferred by usage.
+Numba also supports ``dict(iterable)`` constructor as well.
```
```suggestion
Numba also supports, explicitly, the ``dict(iterable)`` constructor.
```

stuartarchibald commented on numba/tests/test_dictobject.py:
```diff
@@ -1186,6 +1186,134 @@ def test_str(self):
         self.check_stringify(str)
 
 
+class DictIterableCtor(TestCase):
+
+    def setUp(self) -> None:
+        self.jit_enabled = None
+
+    def test_iterable_type_constructor(self):
```
This base class will run as a unit test, is that intended?

stuartarchibald commented on numba/tests/test_dicts.py:
```diff
@@ -37,6 +38,30 @@ def foo():
         self.assertEqual(d, {1: 2})
 
+    def test_use_dict_iterable_args(self):
+        # Test dict(*iterable)
```
```suggestion
        # Test dict(iterable)
```

stuartarchibald commented on numba/typed/typeddict.py:
```diff
@@ -117,6 +117,33 @@ def __init__(self, **kwargs):
+                       "constructor is Dict(iterable)")
+                raise TypeError(msg)
+
+            if isinstance(arg, Mapping):
+                for k, v in arg.items():
+                    self.__setitem__(k, v)
```
This behaviour isn't working under JIT?

guilhermeleobas replied:
I'll leave `dict(mapping)` ctor for a future PR and focus exclusively on `dict(iterable)` in this one.

stuartarchibald commented on numba/tests/test_dictobject.py:
```diff
@@ -1186,6 +1186,134 @@ def test_str(self):
+        with self.assertRaisesRegex(TypingError, msg):
+            Dict({1: 2, 'a': 'b'})
+
+
```
Here's some things that either aren't working or could work better:
```python
from numba import njit
import numpy as np

# Valid, won't compile, fails at typing.
@njit
def foo():
    return dict(([1, 2], [1, 2]))

print(foo())


#Wrong error message, the issue is the types in the iterable aren't supported
#whereas the error message complains of: `Unsupported use of `dict()` with
#positional or keyword argument(s). The only supported uses are `dict()` or
#`dict(*iterable)`.`
@njit
def foo():
    return dict(((1, 2), (3, 'a')))

print(foo())


#Lowering error, should be caught earlier
@njit
def foo():
    return dict((('a', 'b', 'c'), ('d', 'e', 'f')))

print(foo())


#This should really complain about unhashable key type
@njit
def foo():
    return dict((('a', 1), ({}, 2)))

print(foo())


# Mapping support not working under JIT
@njit
def foo():
    return dict({x: x ** 2 for x in range(3)})

print(foo())
```

guilhermeleobas replied:
I believe I have address all the issues pointed out in our comment, except for the last one, which involves `dict(mapping)`.

stuartarchibald reviewed (commented):
Thanks for the patch @guilhermeleobas, I've not looked at the code in depth yet, but have provided some feedback about things that might need looking at in terms of usability and correctness. Thanks again!

guilhermeleobas commented:
Hi @stuartarchibald, I believe I have addressed all your comments. The current failure is not related to any changes introduced in my previous commit.

```
- Windows py310_np122

(testenv) D:\a\1\s>cmd /C conda install -q -y -c numba tbb=2021 tbb-devel 
Collecting package metadata (current_repodata.json): ...working... failed

CondaHTTPError: HTTP 504 GATEWAY TIMEOUT for url <[https://conda.anaconda.org/numba/win-64/repodata.json>](https://conda.anaconda.org/numba/win-64/repodata.json%3E)
Elapsed: 01:00.076791
CF-RAY: 73f5e5f559ed2ad6-ORD

A remote server error occurred when trying to retrieve this URL.
```

esc commented:
@guilhermeleobas @kc611 -- may kindly request an update on the state of this PR?

guilhermeleobas commented:
Hi @esc, PR is waiting for another review

kc611 commented on numba/core/typing/dictdecl.py:
```diff
@@ -17,13 +17,39 @@
+            iterable, = args
+            if isinstance(iterable, types.IterableType):
+                dtype = iterable.iterator_type.yield_type
+                if isinstance(dtype, types.UniTuple):
```
One question:

```python
@njit
def foo():
    return dict(([1, 2], [1, 2])) # Error

print(foo())
```

This is valid within CPython, But this PR only intends to add support when arguments are a list or a tuple of tuples (and not the other cases). Is that correct ?

guilhermeleobas replied:
If possible, I'd like to leave this use case for a future PR.

kc611 reviewed (approved):
I gave it a review, its LGTM (so far). A question below.
